<a href="https://colab.research.google.com/github/Ken-mg/smart-finance-assistant/blob/main/starter_notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 🏦 Project Overview

Welcome to your **Smart Finance Assistant** development journey! This notebook will evolve from basic CSV processing to a complete AI-powered finance application.

**Final Application Components:**
- 💬 **AI Chat Interface** - Financial advice personality
- 📊 **Data Analysis** - CSV transaction processing  
- 🔍 **RAG System** - Retrieval from financial documents
- 🛠️ **Custom Tools** - Calculators and utilities
- 🌐 **Gradio UI** - Professional web interface

**Development Approach**: Build progressively from foundation to advanced features, using AI collaboration throughout.

---

# 🚀 Getting Started: Foundation Setup

## Initial Setup
This cell installs the necessary libraries. In a Colab environment, you would uncomment the first line.

In [17]:
# Import core libraries
import pandas as pd
import numpy as np
from datetime import datetime
import warnings
warnings.filterwarnings('ignore')
!pip install hands-on-ai
!pip install matplotlib-venn
# https://pypi.python.org/pypi/libarchive
!apt-get -qq install -y libarchive-dev && pip install -U libarchive
import libarchive
# https://pypi.python.org/pypi/pydot
!apt-get -qq install -y graphviz && pip install pydot
import pydot

!pip install cartopy
import cartopy



print("📦 Core libraries loaded successfully!")
print(f"Pandas version: {pd.__version__}")

📦 Core libraries loaded successfully!
Pandas version: 2.2.2


## Hands-on-AI Configuration

Set up the hands-on-ai package for advanced features (chat, RAG, tools):

In [18]:
import os

# Configure hands-on-ai server connection
os.environ['HANDS_ON_AI_SERVER'] = ' https://ollama.locallm.org'
os.environ['HANDS_ON_AI_MODEL'] = 'gemma3:4b'
os.environ['HANDS_ON_AI_API_KEY'] = 'Curtin2026ISYS20015002'

print("🔑 Hands-on-AI configured successfully!")

🔑 Hands-on-AI configured successfully!


The API key this semester is:  **isys2001-assignment-key**

## Connection Test

Test that everything is working correctly:

In [19]:
from hands_on_ai.chat import get_response

# Test the connection to the hands-on-ai server
try:
    response = get_response("Hello! I'm building a Smart Finance Assistant.")

    # Check if the response contains an error message instead of blindly trusting it
    if "Error" in response or "Connection error" in response:
        print("❌ Hands-on-AI connection failed (Server returned an error).")
        print(f"Details: {response}")
        print("💡 Tip: You can still work on the data processing foundation without this.")
    else:
        print("✅ Hands-on-AI connection successful!")
        print(f"Response: {response}")

except Exception as e:
    print(f"❌ Hard connection issue/Crash: {e}")


[WARNING] Error during request (attempt 1): Connection error.


Hang tight, I'm thinking... trying again!


[WARNING] Error during request (attempt 2): Connection error.


❌ Hands-on-AI connection failed (Server returned an error).
Details: ❌ Error: Connection error.
💡 Tip: You can still work on the data processing foundation without this.


# 🏗️ Foundation: Data Processing Skills

Before building advanced features, establish solid data processing foundations. This section focuses on CSV transaction analysis - the core of your finance assistant.

## Foundation Skill Checkpoint ✅

**Master these basics before advancing to chat/RAG/tools:**
- [ ] Load and clean CSV transaction data
- [ ] Handle real-world data issues (dollar signs, missing values)
- [ ] Calculate spending summaries by category  
- [ ] Generate business-appropriate insights
- [ ] Format output for professional presentation
- [ ] Test functions with various data scenarios

::: {.callout-tip}
## 🤖 AI Collaboration Strategy

For this foundation work, use AI to:
1. **Generate initial code** with specific business context
2. **Handle data cleaning** and validation
3. **Create professional formatting** for outputs
4. **Suggest business insights** from data patterns
5. **Help with testing** edge cases and error handling

**Remember**: You're directing AI like a junior developer - always review and improve their suggestions!
:::

## Sample Transaction Data Setup

Create or load sample transaction data to work with:

In [20]:
# 🤖 AI Collaboration Opportunity:
# Ask AI to help you create realistic sample transaction data
# Include Australian businesses and various spending categories

# Sample data structure for testing
import pandas as pd
url= "https://raw.githubusercontent.com/Ken-mg/smart-finance-assistant/main/sample_transactions.csv"


# Create DataFrame from sample data
df_sample = pd.read_csv(url)
print("📋 Sample transaction data created:")
print(df_sample)

📋 Sample transaction data created:
          Date   Amount       Category                         Description
0   2024-08-01   $45.50      Groceries              Woolworths Weekly Shop
1   2024-08-02   $12.00      Transport                    Opal Card Top-up
2   2024-08-03   $89.95  Entertainment    Concert Tickets - Enmore Theatre
3   2024-08-04    $3.50         Coffee              Campus Cafe Flat White
4   2024-08-05  $120.00      Groceries                   Coles Weekly Shop
5   2024-08-06  -$25.00         Refund  Returned Textbook - Co-op Bookshop
6   2024-08-07   $85.00         Dining         Birthday Dinner - The Rocks
7   2024-08-08   $15.95         Coffee               Starbucks Double Shot
8   2024-08-09   $67.80      Utilities    Electricity Bill - Origin Energy
9   2024-08-10    $4.20         Coffee              Campus Cafe Cappuccino
10  2024-08-11   $32.50      Transport                     Uber to Airport
11  2024-08-12  $156.00      Groceries         Woolworths Fortnig

---

# 📊 Six-Step Development Methodology

Your notebook must demonstrate the six-step methodology with clear evidence of AI collaboration at each step.

## STEP 1: Understand the Problem

**🎯 Define Your Finance Problem**

In this section, clearly state your chosen finance problem in business terms.

::: {.callout-note}
## Problem Definition Template

*Example: "I want to help people understand their spending habits by analyzing their bank transaction exports. The system should identify spending patterns, highlight potential savings opportunities, and provide actionable financial insights in easy-to-understand language."*

**🤖 AI Prompt to Try:**
```
"Help me brainstorm specific personal finance problems that could be solved
with a CSV analysis tool. I'm interested in helping people with budgeting
and spending awareness. What are common financial challenges that data
analysis could address?"
```

**Your Problem Statement:**
[Write your specific finance problem here - be clear about who you're helping and what value you're providing]

## STEP 2: Identify Inputs and Outputs

**📥 Define Your Data Flow**

Clearly specify what data you'll work with and what insights you'll generate.

## Input/Output Analysis Template

**Inputs:**
- CSV file with transaction data (columns: Date, Amount, Category, Description)
- User preferences (spending goals, categories to focus on)
- Time period for analysis

**Outputs:**
- Total spending summary by category
- Average spending patterns
- Spending trend analysis
- Actionable savings recommendations
- Visual spending breakdown (future enhancement)

**🤖 AI Prompt to Try:**
```
"For a personal finance assistant that analyzes CSV transaction data,
what are the most valuable inputs I should collect and outputs I should
provide? Consider both technical data processing and business value
for the user."
```

**Your Input/Output Definition:**
[Define your specific inputs and outputs here]

## STEP 3: Work the Problem by Hand

**✋ Manual Calculation Examples**

Show 2-3 worked examples to understand the logic before coding.


## Example Business Calculation

Given this sample data:

| Date | Amount | Category | Description |
|------|--------|----------|-------------|
| 2024-08-01 | $45.50 | Groceries | Woolworths |
| 2024-08-02 | $12.00 | Transport | Opal Card |
| 2024-08-03 | $89.95 | Entertainment | Concert |
| 2024-08-04 | -$15.00 | Refund | Returned item |

**Manual Calculations:**
- Total Spending: $45.50 + $12.00 + $89.95 - $15.00 = $132.45
- By Category: Groceries $45.50, Transport $12.00, Entertainment $89.95
- Average Transaction: $132.45 ÷ 4 = $33.11
- Insight: Entertainment represents 67% of positive spending

**🤖 AI Prompt to Try:**
```
"Give me 3 realistic examples of Australian transaction data with different
scenarios (normal spending, refunds, large purchases). Show me how to manually
calculate meaningful financial insights from each scenario."
```

**Your Manual Examples:**
[Work through your own examples here - this helps you understand the business logic]

## STEP 4: Write Pseudocode

**📝 Plan Your Solution Logic**

Sketch the algorithm in plain English before coding.


## Pseudocode Template

```
FUNCTION analyze_spending_data(csv_file):
    // Step 1: Load and validate data
    - Read CSV file into DataFrame
    - Check for required columns (Date, Amount, Category, Description)
    - Handle missing or invalid data
    
    // Step 2: Clean financial data  
    - Remove dollar signs from Amount column
    - Convert to numeric values
    - Separate positive spending from refunds
    
    // Step 3: Calculate business metrics
    - Group transactions by category
    - Calculate totals, averages, counts per category
    - Identify spending patterns and outliers
    
    // Step 4: Generate insights
    - Calculate percentages of total spending
    - Identify top spending categories
    - Suggest potential savings opportunities
    
    // Step 5: Format for presentation
    - Create business-friendly output
    - Include currency formatting
    - Add explanatory text for insights
    
    RETURN formatted_analysis
```

**🤖 AI Prompt to Try:**
```
"Review my pseudocode for analyzing spending data from a CSV. Are there any
edge cases I'm missing? What business logic should I add to make this more
valuable for personal finance insights?"
```


**Your Pseudocode:**
[Write your detailed algorithm here]

## STEP 5: Convert to Python

**💻 Implementation with AI Collaboration**

Now implement your solution using AI assistance. Focus on creating professional, business-appropriate code.


## 🤖 Implementation Strategy

**Effective AI Prompts for Implementation:**
```
"I'm implementing a Smart Finance Assistant. Based on my pseudocode, please create
a Python function that [specific functionality]. The code should:
- Handle real-world CSV data issues (dollar signs, missing values)
- Include clear comments explaining business logic
- Use professional variable names
- Format output for business presentation
- Include basic error handling"
```

**Remember to critique and improve AI responses before using them!**
:::

### Foundation Data Processing Functions

In [35]:
def load_and_clean_transaction_data(file_path):
    """
    Load and clean transaction data for the Smart Finance Assistant
    """
    import pandas as pd
    import io

    try:
        # Load data from file path or file object
        if isinstance(file_path, str):
            df = pd.read_csv(file_path)
        else:
            df = pd.read_csv(io.StringIO(file_path.read().decode("utf-8")))

        # Validate required columns
        required_columns = ["Date", "Amount", "Category", "Description"]
        missing = [col for col in required_columns if col not in df.columns]
        if missing:
            raise ValueError(f"Missing required columns: {missing}")

        # Clean Amount column: remove dollar signs, commas, convert to float
        df["Amount"] = (
            df["Amount"]
            .astype(str)
            .str.replace(r"[\$,]", "", regex=True)
            .str.strip()
        )
        df["Amount"] = pd.to_numeric(df["Amount"], errors="coerce")

        # Drop rows where Amount couldn't be parsed
        invalid_rows = df["Amount"].isna().sum()
        if invalid_rows > 0:
            print(f"⚠️ Warning: {invalid_rows} row(s) with invalid amounts were removed.")
        df = df.dropna(subset=["Amount"])

        # Parse dates
        df["Date"] = pd.to_datetime(df["Date"], errors="coerce")

        # Fill missing categories/descriptions
        df["Category"] = df["Category"].fillna("Uncategorized")
        df["Description"] = df["Description"].fillna("No description")

        print(f"✅ Successfully loaded {len(df)} transactions.")
        return df

    except FileNotFoundError:
        print(f"❌ Error: File '{file_path}' not found.")
        return None
    except Exception as e:
        print(f"❌ Unexpected error: {e}")
        return None
# Test
url = "https://raw.githubusercontent.com/Ken-mg/smart-finance-assistant/main/sample_transactions.csv"
clean_data = load_and_clean_transaction_data(url)
print(clean_data)

✅ Successfully loaded 25 transactions.
         Date  Amount       Category                         Description
0  2024-08-01   45.50      Groceries              Woolworths Weekly Shop
1  2024-08-02   12.00      Transport                    Opal Card Top-up
2  2024-08-03   89.95  Entertainment    Concert Tickets - Enmore Theatre
3  2024-08-04    3.50         Coffee              Campus Cafe Flat White
4  2024-08-05  120.00      Groceries                   Coles Weekly Shop
5  2024-08-06  -25.00         Refund  Returned Textbook - Co-op Bookshop
6  2024-08-07   85.00         Dining         Birthday Dinner - The Rocks
7  2024-08-08   15.95         Coffee               Starbucks Double Shot
8  2024-08-09   67.80      Utilities    Electricity Bill - Origin Energy
9  2024-08-10    4.20         Coffee              Campus Cafe Cappuccino
10 2024-08-11   32.50      Transport                     Uber to Airport
11 2024-08-12  156.00      Groceries         Woolworths Fortnightly Shop
12 2024-08-1

In [32]:
# 🤖 AI Collaboration: Spending Analysis Function
# Ask AI to help you create comprehensive spending analysis
def analyze_spending_patterns(df):
    """
    Analyze spending patterns and generate business insights
    """
    import pandas as pd

    # Distinguish between expenses (positive) and refunds (negative)

    spending_df = df[df["Amount"] > 0].copy()
    total_spending = spending_df["Amount"].sum()

    # Analyzing per category
    category_analysis = (
        spending_df.groupby("Category")["Amount"]
        .agg(["sum", "count", "mean"])
        .rename(columns={"sum": "Total", "count": "Transactions", "mean": "Average"})
        .sort_values("Total", ascending=False)
    )

    # percentage adding
    category_analysis["Percentage"] = (
        (category_analysis["Total"] / total_spending * 100).round(2)
    )

    # --- Identifying les spending areas ---
    top_category = category_analysis.index[0]
    top_amount = category_analysis["Total"].iloc[0]
    top_pct = category_analysis["Percentage"].iloc[0]

    high_spend_areas = category_analysis[
        category_analysis["Percentage"] > 30
    ].index.tolist()

    # --- Business Insights ---
    insights = []

    insights.append(
        f"💰 Total Spending: ${total_spending:,.2f} across "
        f"{len(spending_df)} transactions."
    )
    insights.append(
        f"📊 Top Spending Category: '{top_category}' represents "
        f"{top_pct}% of your total budget (${top_amount:,.2f})."
    )
    if high_spend_areas:
        areas = ", ".join(high_spend_areas)
        insights.append(
            f"⚠️  High Spending Alert: {areas} exceed 30% of your budget. "
            f"Consider reviewing these areas for potential savings."
        )

    highest_avg_cat = category_analysis["Average"].idxmax()
    highest_avg_val = category_analysis["Average"].max()
    insights.append(
        f"🔍 Highest Average Transaction: '{highest_avg_cat}' averages "
        f"${highest_avg_val:,.2f} per transaction — review for discretionary cuts."
    )

    if top_pct > 50:
        insights.append(
            f"💡 Recommendation: Over half your spending goes to '{top_category}'. "
            f"Setting a monthly cap could improve your financial balance."
        )
    else:
        insights.append(
            "✅ Your spending appears relatively balanced across categories. "
            "Focus on maintaining this distribution."
        )

    print("=" * 55)
    print("       SMART FINANCE ASSISTANT — SPENDING REPORT")
    print("=" * 55)

    print(f"\n📋 SPENDING BY CATEGORY\n{'-' * 55}")
    for cat, row in category_analysis.iterrows():
        bar = "█" * int(row["Percentage"] / 5)
        print(
            f"  {cat:<20} ${row['Total']:>8,.2f}   "
            f"{row['Percentage']:>5.1f}%  {bar}"
        )

    print(f"\n🔎 BUSINESS INSIGHTS\n{'-' * 55}")
    for insight in insights:
        print(f"  {insight}")

    print("=" * 55)

    return {
        "total_spending": total_spending,
        "category_analysis": category_analysis,
        "top_category": top_category,
        "high_spend_areas": high_spend_areas,
        "insights": insights,
    }

# Test
analysis = analyze_spending_patterns(clean_data)

       SMART FINANCE ASSISTANT — SPENDING REPORT

📋 SPENDING BY CATEGORY
-------------------------------------------------------
  Groceries            $  540.30    44.7%  ████████
  Entertainment        $  206.95    17.1%  ███
  Utilities            $  159.80    13.2%  ██
  Dining               $  148.85    12.3%  ██
  Transport            $  108.00     8.9%  █
  Coffee               $   44.75     3.7%  

🔎 BUSINESS INSIGHTS
-------------------------------------------------------
  💰 Total Spending: $1,208.65 across 23 transactions.
  📊 Top Spending Category: 'Groceries' represents 44.7% of your total budget ($540.30).
  ⚠️  High Spending Alert: Groceries exceed 30% of your budget. Consider reviewing these areas for potential savings.
  🔍 Highest Average Transaction: 'Groceries' averages $108.06 per transaction — review for discretionary cuts.
  ✅ Your spending appears relatively balanced across categories. Focus on maintaining this distribution.


In [33]:
def generate_financial_recommendations(analysis_data):
    """
    Based on spending analysis data, create professional financial
    recommendations. Include specific savings opportunities, spending
    pattern observations, and actionable advice formatted for a
    personal finance app user.

    Args:
        analysis_data: Dictionary with spending analysis results
    """

    category_analysis = analysis_data["category_analysis"]
    total_spending = analysis_data["total_spending"]
    top_category = analysis_data["top_category"]
    high_spend_areas = analysis_data["high_spend_areas"]

    recommendations = []
    savings_opportunities = []

    # --- Spending Pattern Observations ---
    observations = []
    for cat, row in category_analysis.iterrows():
        if row["Percentage"] > 30:
            observations.append(
                f"📌 '{cat}' takes up {row['Percentage']}% of your budget — "
                f"this is significantly high."
            )
        elif row["Percentage"] > 15:
            observations.append(
                f"📌 '{cat}' represents {row['Percentage']}% of your spending — "
                f"worth monitoring."
            )

    # --- Savings Opportunities ---
    for cat, row in category_analysis.iterrows():
        if row["Percentage"] > 30:
            potential_saving = row["Total"] * 0.20
            savings_opportunities.append(
                f"💡 Reducing '{cat}' spending by 20% could save you "
                f"${potential_saving:,.2f} per month."
            )
        if row["Average"] > 100:
            savings_opportunities.append(
                f"💡 Your average '{cat}' transaction is ${row['Average']:,.2f} — "
                f"consider cheaper alternatives."
            )

    # --- Actionable Recommendations ---
    recommendations.append(
        f"🎯 Set a monthly budget cap for '{top_category}' — "
        f"it is your biggest expense at ${category_analysis.loc[top_category, 'Total']:,.2f}."
    )
    recommendations.append(
        "🎯 Review your subscriptions and recurring payments — "
        "cancel any services you no longer use."
    )
    recommendations.append(
        f"🎯 Try the 50/30/20 rule: 50% needs, 30% wants, 20% savings. "
        f"Your current total spending is ${total_spending:,.2f}."
    )
    if high_spend_areas:
        recommendations.append(
            f"🎯 Focus on reducing spending in: {', '.join(high_spend_areas)}. "
            f"These areas exceed 30% of your budget."
        )

    # --- Formatted Output ---
    print("=" * 55)
    print("     SMART FINANCE ASSISTANT — RECOMMENDATIONS")
    print("=" * 55)

    print(f"\n🔍 SPENDING PATTERN OBSERVATIONS\n{'-' * 55}")
    if observations:
        for obs in observations:
            print(f"  {obs}")
    else:
        print("  ✅ No unusual spending patterns detected.")

    print(f"\n💰 SAVINGS OPPORTUNITIES\n{'-' * 55}")
    if savings_opportunities:
        for saving in savings_opportunities:
            print(f"  {saving}")
    else:
        print("  ✅ Your spending looks optimized!")

    print(f"\n✅ ACTIONABLE RECOMMENDATIONS\n{'-' * 55}")
    for rec in recommendations:
        print(f"  {rec}")

    print("=" * 55)

    return {
        "observations": observations,
        "savings_opportunities": savings_opportunities,
        "recommendations": recommendations
    }

# Test
recommendations = generate_financial_recommendations(analysis)

     SMART FINANCE ASSISTANT — RECOMMENDATIONS

🔍 SPENDING PATTERN OBSERVATIONS
-------------------------------------------------------
  📌 'Groceries' takes up 44.7% of your budget — this is significantly high.
  📌 'Entertainment' represents 17.12% of your spending — worth monitoring.

💰 SAVINGS OPPORTUNITIES
-------------------------------------------------------
  💡 Reducing 'Groceries' spending by 20% could save you $108.06 per month.
  💡 Your average 'Groceries' transaction is $108.06 — consider cheaper alternatives.

✅ ACTIONABLE RECOMMENDATIONS
-------------------------------------------------------
  🎯 Set a monthly budget cap for 'Groceries' — it is your biggest expense at $540.30.
  🎯 Review your subscriptions and recurring payments — cancel any services you no longer use.
  🎯 Try the 50/30/20 rule: 50% needs, 30% wants, 20% savings. Your current total spending is $1,208.65.
  🎯 Focus on reducing spending in: Groceries. These areas exceed 30% of your budget.


---

# 🌐 Advanced Features: Integrating AI Components

Once your foundation data processing is solid, integrate advanced AI features using hands-on-ai.

## Chat Interface Integration

In [24]:
headers = {
    "Content-Type": "application/json",
    "x-api-key": "Curtin2026ISYS20015002",        # ← just this line
    "anthropic-version": "2023-06-01"
}
def get_response(message, system=None):
    import urllib.request, json

    headers = {"Content-Type": "application/json"}
    body = {
        "model": "claude-sonnet-4-20250514",
        "max_tokens": 1000,
        "messages": [{"role": "user", "content": message}]
    }
    if system:
        body["system"] = system

    req = urllib.request.Request(
        "https://api.anthropic.com/v1/messages",
        data=json.dumps(body).encode(),
        headers=headers,
        method="POST"
    )
    with urllib.request.urlopen(req) as resp:
        data = json.loads(resp.read())
        return data["content"][0]["text"]
def create_finance_chatbot_system_prompt():
    """
    Creates a system prompt for a friendly, professional financial
    advisor chatbot that can provide spending advice based on transaction
    analysis. The personality should be encouraging but practical.
    """
    system_prompt = """
    You are FinBot, a friendly and professional personal finance advisor assistant.

    YOUR PERSONALITY:
    - Encouraging and supportive, never judgmental about spending habits
    - Practical and actionable — always give concrete next steps
    - Clear and simple — avoid financial jargon, speak like a trusted friend
    - Positive but honest — celebrate wins while addressing areas to improve

    YOUR EXPERTISE:
    - Personal budgeting and expense tracking
    - Spending pattern analysis and categorization
    - Savings strategies and goal setting
    - Everyday financial decisions (groceries, transport, entertainment)

    YOUR RESPONSE STYLE:
    - Keep responses concise and easy to read
    - Use bullet points for actionable advice
    - Always end with an encouraging note or next step
    - Use simple $ amounts and percentages when relevant

    IMPORTANT RULES:
    - Never provide investment or stock market advice
    - Never ask for sensitive personal information
    - Always remind users you are an AI, not a certified financial advisor
    - Focus on practical day-to-day financial habits
    """
    return system_prompt


def chat_with_finance_bot(user_message, analysis_data=None):
    """
    Chat with the finance bot using hands_on_ai.
    Optionally pass in spending analysis data for personalised advice.

    Args:
        user_message  : The user's question or message
        analysis_data : Optional dict from analyze_spending_patterns()
    Returns:
        str: The chatbot's response
    """
    from hands_on_ai.chat import get_response

    system_prompt = create_finance_chatbot_system_prompt()

    # If analysis data is available, add context to the message
    if analysis_data:
        top_category = analysis_data["top_category"]
        total_spending = analysis_data["total_spending"]
        high_spend_areas = ", ".join(analysis_data["high_spend_areas"]) or "None"

        context = (
            f"\n\n[USER SPENDING CONTEXT]\n"
            f"Total Spending: ${total_spending:,.2f}\n"
            f"Top Category: {top_category}\n"
            f"High Spend Areas: {high_spend_areas}\n"
        )
        full_message = user_message + context
    else:
        full_message = user_message

    response = get_response(full_message, system=system_prompt)
    return response


# Test your chatbot
questions = [
    "I spend too much on coffee, what should I do?",
    "How can I start saving more money each month?",
    "What is the 50/30/20 budgeting rule?",
]

print("=" * 60)
print("           FINBOT — YOUR PERSONAL FINANCE ADVISOR")
print("=" * 60)

for question in questions:
    print(f"\n🧑 You: {question}")
    response = chat_with_finance_bot(question, analysis_data=analysis)
    print(f"🤖 FinBot: {response}")
    print("-" * 60)

           FINBOT — YOUR PERSONAL FINANCE ADVISOR

🧑 You: I spend too much on coffee, what should I do?


[WARNING] Error during request (attempt 1): Connection error.


Oops! I got a little tangled up... Let's try that again 😊


[WARNING] Error during request (attempt 2): Connection error.


🤖 FinBot: ❌ Error: Connection error.
------------------------------------------------------------

🧑 You: How can I start saving more money each month?


[WARNING] Error during request (attempt 1): Connection error.


Hang tight, I'm thinking... trying again!


[WARNING] Error during request (attempt 2): Connection error.


🤖 FinBot: ❌ Error: Connection error.
------------------------------------------------------------

🧑 You: What is the 50/30/20 budgeting rule?


[WARNING] Error during request (attempt 1): Connection error.


Hang tight, I'm thinking... trying again!


[WARNING] Error during request (attempt 2): Connection error.


🤖 FinBot: ❌ Error: Connection error.
------------------------------------------------------------


## RAG System for Financial Documents

In [25]:
# --- SIMULATION DU SYSTÈME RAG (Au cas où hands_on_ai n'est pas installé) ---
# Si tu as le vrai dossier "hands_on_ai", tu peux décommenter la ligne suivante :
# from hands_on_ai.rag import RAG

class RAG:
    """Une classe de simulation pour que ton code fonctionne immédiatement"""
    def __init__(self):
        self.docs = []
    def ingest(self, documents):
        self.docs = documents
    def ask(self, question):
        # Simulation d'une recherche basique par mot-clé
        for doc in self.docs:
            for word in question.split():
                if len(word) > 4 and word.lower() in doc.lower():
                    return doc
        return "Je n'ai pas trouvé de réponse exacte dans mes documents, mais je vous conseille de rester prudent avec vos dépenses."

# --- INSTALLATIONS REQUISES ---
!pip install -q matplotlib-venn cartopy pydot
!apt-get -qq install -y libfluidsynth1 libarchive-dev graphviz
!pip install -q -U libarchive

import libarchive
import pydot
import cartopy

# --- FONCTIONS DU SYSTÈME FINANCIAL RAG ---

def setup_financial_rag():
    """
    Set up RAG system for financial documents and transaction data.
    """
    rag_system = RAG()

    # --- Financial knowledge documents ---
    documents = [
        "The 50/30/20 rule suggests spending 50% of income on needs, "
        "30% on wants, and saving 20%. Needs include rent, groceries, "
        "and transport. Wants include entertainment and dining out.",

        "Zero-based budgeting means assigning every dollar of income "
        "a purpose. At the end of the month, income minus expenses "
        "should equal zero. This method encourages intentional spending.",

        "An emergency fund should cover 3 to 6 months of living expenses. "
        "Keep it in a separate savings account so you are not tempted "
        "to spend it on non-emergencies.",

        "If you overspend on entertainment, try setting a monthly cap. "
        "Use free alternatives like parks, libraries, and community events. "
        "Cancel unused subscriptions to streaming or gaming services.",

        "Food and dining out is one of the most common overspending areas. "
        "Meal prepping on weekends, making a grocery list, and avoiding "
        "impulse purchases can significantly reduce food expenses.",

        "Transport costs can be reduced by using public transport, "
        "carpooling, or cycling for short distances. Review whether "
        "owning a car is necessary if you live near public transit.",

        "Automating savings means setting up an automatic transfer to "
        "your savings account on payday. This ensures you save before "
        "you spend, making it easier to reach financial goals.",

        "Tracking daily expenses helps identify spending leaks. "
        "Small purchases like coffee, snacks, and impulse buys "
        "can add up to hundreds of dollars per month.",

        "Paying yourself first is a strategy where savings are treated "
        "as a non-negotiable expense. Set aside at least 10 to 20 percent "
        "of your income before spending on anything else.",

        "High interest debt like credit cards should be paid off as "
        "quickly as possible. The avalanche method targets the highest "
        "interest debt first, saving the most money over time.",

        "The snowball method focuses on paying off the smallest debt "
        "first to build momentum and motivation. Once a debt is cleared, "
        "redirect that payment to the next smallest debt.",
    ]

    rag_system.ingest(documents)
    print(f"✅ RAG system ready — {len(documents)} financial documents loaded.")
    return rag_system


def ask_finance_rag(rag_system_instance, question, analysis_data=None):
    """
    Ask the RAG system a personal finance question.
    """
    if analysis_data:
        top_category = analysis_data["top_category"]
        total = analysis_data["total_spending"]
        context = (
            f" My top spending category is {top_category} "
            f"and my total spending is ${total:,.2f}."
        )
        enriched_question = question + context
    else:
        enriched_question = question

    answer = rag_system_instance.ask(enriched_question)
    return answer


# --- SETUP ET TEST ---

# 1. Définition des données d'analyse manquantes pour éviter l'erreur
analysis = {
    "top_category": "Entertainment",
    "total_spending": 450.00
}

# 2. Initialisation du RAG
financial_rag_system = setup_financial_rag()

print("\n" + "=" * 60)
print("        SMART FINANCE ASSISTANT — RAG Q&A")
print("=" * 60)

test_questions = [
    "What's a good budgeting strategy for someone who overspends on entertainment?",
    "How much should I have in my emergency fund?",
    "What is the 50/30/20 rule?",
]

for question in test_questions:
    print(f"\n🧑 Question : {question}")
    # Utilisation de la variable 'analysis' définie plus haut
    answer = ask_finance_rag(financial_rag_system, question, analysis_data=analysis)
    print(f"🤖 Answer   : {answer}")
    print("-" * 60)

E: Package 'libfluidsynth1' has no installation candidate
✅ RAG system ready — 11 financial documents loaded.

        SMART FINANCE ASSISTANT — RAG Q&A

🧑 Question : What's a good budgeting strategy for someone who overspends on entertainment?
🤖 Answer   : The 50/30/20 rule suggests spending 50% of income on needs, 30% on wants, and saving 20%. Needs include rent, groceries, and transport. Wants include entertainment and dining out.
------------------------------------------------------------

🧑 Question : How much should I have in my emergency fund?
🤖 Answer   : The 50/30/20 rule suggests spending 50% of income on needs, 30% on wants, and saving 20%. Needs include rent, groceries, and transport. Wants include entertainment and dining out.
------------------------------------------------------------

🧑 Question : What is the 50/30/20 rule?
🤖 Answer   : The 50/30/20 rule suggests spending 50% of income on needs, 30% on wants, and saving 20%. Needs include rent, groceries, and transport

In [26]:
# https://pypi.python.org/pypi/libarchive
!apt-get -qq install -y libarchive-dev && pip install -U libarchive
import libarchive

## Custom Financial Tools

In [27]:
# 🤖 AI Collaboration: Custom Tool Development
# Ask AI to help you create useful financial calculators

from hands_on_ai import agent

def create_savings_calculator_tool():
    """
    Create a custom savings goal calculator as an agent tool

    🤖 AI Collaboration Prompt:
    "Create a savings goal calculator function that takes current savings,
    monthly contribution, and target amount, then calculates time to reach
    goal. Format output for user-friendly display."
    """
    # Your AI-assisted tool implementation goes here
    # Include input validation, calculations, formatted output
    pass

# Register your tool with the agent system
# agent.register_tool("savings_calculator", create_savings_calculator_tool)

## Gradio UI Integration

In [28]:
# 🤖 AI Collaboration: Professional UI Design
# Ask AI to help you create a comprehensive Gradio interface

import gradio as gr

def create_finance_assistant_ui():
    """
    Create a comprehensive Gradio interface for the Smart Finance Assistant

    🤖 AI Collaboration Prompt:
    "Help me design a Gradio interface that combines CSV upload, spending
    analysis, chat functionality, and custom tools in a user-friendly
    layout suitable for a personal finance application."
    """
    # Your AI-assisted UI implementation goes here
    # Include multiple tabs, file uploads, chat interface, tool access
    pass

# Launch your complete application
# demo = create_finance_assistant_ui()
# demo.launch()

---

# 🧪 STEP 6: Test with a Variety of Data

**🔍 Comprehensive Testing Strategy**

Create thorough tests for your Smart Finance Assistant to ensure it handles real-world scenarios.

::: {.callout-tip}
## 🤖 AI Collaboration for Testing

**Effective Testing Prompts:**
```
"Help me create comprehensive test cases for my finance assistant. Include:
- Normal transaction data
- Edge cases (refunds, large amounts, missing data)
- Invalid data scenarios (corrupted files, wrong formats)
- Business logic validation (spending calculations, recommendations)
Create assert statements to verify each scenario."
```
:::

## Foundation Function Tests

In [29]:
# 🤖 AI Collaboration: Comprehensive Test Suite
# Ask AI to help you create thorough test cases

def create_test_datasets():
    """
    Create various test datasets for comprehensive testing

    🤖 AI Collaboration Prompt:
    "Create realistic test datasets for a finance assistant including:
    1. Normal spending data with various categories
    2. Edge cases: refunds (negative amounts), missing data, zero amounts
    3. Data quality issues: invalid formats, extreme values
    4. Business scenarios: high spending months, savings patterns
    Include Australian business names and realistic amounts."
    """
    # Your AI-generated test data goes here
    pass

def test_data_loading_function():
    """
    Test the data loading and cleaning functionality

    🤖 AI Collaboration Prompt:
    "Create assert statements to test my data loading function with:
    - Valid CSV data
    - CSV with dollar signs in amounts
    - Missing values and invalid data
    - Empty files and corrupted data
    Verify that cleaning works correctly and errors are handled gracefully."
    """
    print("🧪 Testing data loading function...")
    # Your AI-generated test cases go here
    pass

def test_spending_analysis():
    """
    Test spending analysis calculations

    🤖 AI Collaboration Prompt:
    "Create tests for spending analysis that verify:
    - Category totals are calculated correctly
    - Percentages add up to 100%
    - Refunds are handled appropriately
    - Edge cases like single transactions or empty categories
    Use assert statements with known expected results."
    """
    print("🧪 Testing spending analysis...")
    # Your AI-generated analysis tests go here
    pass

def test_business_insights():
    """
    Test business recommendation generation

    🤖 AI Collaboration Prompt:
    "Create tests that verify business insights are appropriate:
    - High spending categories are identified correctly
    - Savings opportunities are realistic
    - Recommendations match spending patterns
    - Output format is user-friendly"
    """
    print("🧪 Testing business insights...")
    # Your AI-generated insight tests go here
    pass

# Run all tests
print("🔍 COMPREHENSIVE TESTING SUITE")
print("=" * 40)

try:
    create_test_datasets()
    test_data_loading_function()
    test_spending_analysis()
    test_business_insights()
    print("✅ All tests passed! Your finance assistant is working correctly.")
except AssertionError as e:
    print(f"❌ Test failed: {e}")
except Exception as e:
    print(f"⚠️ Test error: {e}")

🔍 COMPREHENSIVE TESTING SUITE
🧪 Testing data loading function...
🧪 Testing spending analysis...
🧪 Testing business insights...
✅ All tests passed! Your finance assistant is working correctly.


## Advanced Integration Tests

In [30]:
# 🤖 AI Collaboration: Integration Testing
# Ask AI to help test the complete system integration

def test_full_workflow():
    """
    Test the complete workflow from CSV upload to final recommendations

    🤖 AI Collaboration Prompt:
    "Create an end-to-end test that:
    1. Loads sample CSV data
    2. Runs complete analysis pipeline
    3. Generates chat responses about the data
    4. Verifies RAG system retrieval
    5. Tests custom tool functionality
    Ensure all components work together seamlessly."
    """
    print("🧪 Testing complete workflow integration...")
    # Your AI-generated integration tests go here
    pass

def test_error_handling():
    """
    Test error handling and user experience

    🤖 AI Collaboration Prompt:
    "Create tests that verify error handling for:
    - Invalid file uploads
    - Network connection issues
    - Malformed data
    - User input validation
    Ensure error messages are user-friendly and helpful."
    """
    print("🧪 Testing error handling...")
    # Your AI-generated error tests go here
    pass

# Run integration tests
try:
    test_full_workflow()
    test_error_handling()
    print("✅ Integration tests completed successfully!")
except Exception as e:
    print(f"⚠️ Integration test issue: {e}")

🧪 Testing complete workflow integration...
🧪 Testing error handling...
✅ Integration tests completed successfully!


---

# 📊 Project Completion Checklist

## Foundation Skills ✅
- [ ] **Data Processing**: CSV loading and cleaning functions work reliably
- [ ] **Analysis Functions**: Spending summaries calculate correctly
- [ ] **Business Insights**: Recommendations are relevant and actionable  
- [ ] **Error Handling**: Graceful handling of data issues
- [ ] **Testing**: Comprehensive test coverage for core functions
- [ ] **Documentation**: Clear AI collaboration documentation in diary

## Advanced Integration ✅
- [ ] **Chat Interface**: Finance advisor personality implemented
- [ ] **RAG System**: Document retrieval for financial guidance
- [ ] **Custom Tools**: At least one financial calculator/utility
- [ ] **Gradio UI**: Professional, user-friendly interface
- [ ] **Full Integration**: All components work together seamlessly

## Professional Standards ✅
- [ ] **Code Quality**: Professional, commented, maintainable code
- [ ] **Business Focus**: Clear connection to real finance problems
- [ ] **User Experience**: Interface suitable for non-technical users
- [ ] **AI Collaboration**: Extensive, well-documented AI usage
- [ ] **Testing**: Robust validation of all features

## Project Documentation ✅  
- [ ] **Developer's Diary**: Complete AI collaboration documentation
- [ ] **README**: Clear project description and usage instructions
- [ ] **GitHub**: Regular commits showing development progress
- [ ] **Reflection**: Thoughtful analysis of learning and challenges

---

# 🎯 Final Thoughts: Your Finance Assistant Journey

Congratulations on building your Smart Finance Assistant! This project represents a significant achievement in modern business programming:

**Technical Skills Developed:**
- AI-assisted development workflows
- Professional data processing with pandas
- Integration of multiple AI technologies
- User interface design with Gradio
- Comprehensive software testing

**Business Skills Developed:**  
- Financial data analysis and insights
- User-centered application design
- Professional documentation practices
- Iterative development methodology
- Critical evaluation of AI suggestions

**Professional Preparation:**
- Experience with industry-standard AI collaboration
- Portfolio-ready application development
- Understanding of business problem-solving with technology
- Documentation practices for workplace environments

**Your Smart Finance Assistant demonstrates your ability to direct AI tools toward meaningful business solutions - exactly the skill set that modern BIS graduates need for career success!**

---

*Remember to document all AI collaborations in your Developer's Diary and maintain regular GitHub commits throughout your development process.*
